File: cpp/sfo_45_arrange_an_array_into_the_smallest_number_s2/sfo_45_arrange_an_array_into_the_smallest_number_s2.cpp
/*
* File: sfo_45_arrange_an_array_into_the_smallest_number_s2.cpp
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    string minNumber(vector<int>& nums) {
        vector<string> strs;
        string res;
        for(int i = 0; i < nums.size(); i++)
            strs.push_back(to_string(nums[i]));
        sort(strs.begin(), strs.end(), [](string& x, string& y){ return x + y < y + x; });
        for(int i = 0; i < strs.size(); i++)
            res.append(strs[i]);
        return res;
    }
};

int main() {
    // ======= Test Case =======
    vector<int> nums = { 3, 30, 34, 5, 9 };
    // ====== Driver Code ======
    Solution* slt = new Solution();
    string res = slt->minNumber(nums);
    cout << res << endl;
    
    return 0;
}

File: sword_for_offer/codes/java/sfo_45_arrange_an_array_into_the_smallest_number_s2/sfo_45_arrange_an_array_into_the_smallest_number_s2.java
/*
* File: sfo_45_arrange_an_array_into_the_smallest_number_s2.java
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

package sfo_45_arrange_an_array_into_the_smallest_number_s2;

import include.*;
import java.util.*;

// ===== Solution Code =====
class Solution {
    public String minNumber(int[] nums) {
        String[] strs = new String[nums.length];
        for(int i = 0; i < nums.length; i++)
            strs[i] = String.valueOf(nums[i]);
        Arrays.sort(strs, (x, y) -> (x + y).compareTo(y + x));
        StringBuilder res = new StringBuilder();
        for(String s : strs)
            res.append(s);
        return res.toString();
    }
}

public class sfo_45_arrange_an_array_into_the_smallest_number_s2 {
    public static void main(String[] args) {
        // ======= Test Case =======
        int[] nums = { 3, 30, 34, 5, 9 };
        // ====== Driver Code ======
        Solution slt = new Solution();
        String res = slt.minNumber(nums);
        System.out.println(res);
    }
}
